### PR TITLE
Refactor kv.Store to interface

### PR DESCRIPTION
**Pull request by morlay**
Make `kv.Store` as interface 

We may replace to any other kv storage or distributed/remote kv for https://github.com/genjidb/genji/issues/48

### CODE REVIEW

**Comment by asdine** [comment]
Hey @morlay thanks for submitting this PR and for putting so much effort!
Unfortunately, I can't accept this change because of the following reasons:

- Genji just went from having multiple store implementations (see [v0.14](https://github.com/genjidb/genji/tree/v0.14.1/engine), we had 3 different implementations) to a single one. Maintaining multiple implementations is a LOT of work, I have been trying it for years and I believe is not worth it. [Latest release](https://github.com/genjidb/genji/releases/tag/v0.15.0) focuses on a single backend for exactly this reason.
 - Most kv-stores come with their own transactions, whereas Genji implements its own transaction mechanism which relies a lot on Pebble primitives, which are hard to find in other kv-stores.
- Even if we decide to merge this PR, you will quickly realize how hard it is to implement another kv-store: Most of them don't allow you to pass your own custom [Compare](https://github.com/genjidb/genji/blob/main/internal/encoding/helpers.go#L100) function. Keys in Genji are not lexicographically sorted, but require to be in a certain order when we iterate on them. Most kv-stores use `bytes.Compare` to determine which key is greater than the other, and this won't work with the [binary](https://github.com/genjidb/genji/blob/main/internal/tree/key.go#L26-L47) representation of keys.

Basically, focusing on a single kv-store makes everything *much* easier to maintain, optimize and control.

They are plenty of things to do in Genji, and I would be more than happy to discuss about ideas in issues, so please don't hesitate to open one if you want any change.

**Comment by morlay** [comment]
@asdine 

I just want to extends Genji scalable like the issue metioned https://github.com/genjidb/genji/issues/48.
`kv.Store` as interface could reach this goal, with implememting something like `raft pebble`.

Then Genji will be a Progressive Storage, from embed to distributed 😄 .

**Comment by asdine** [comment]
@morlay I don't think this is the right way to make Genji distributed. This is a too big of a change to be done without proper discussion and planning sorry.
At the moment, the focus is mainly on adding SQL features and reaching stability. Even if we added raft to Genji, it wouldn't be very useful as it doesn't yet have enough SQL features.
I'm closing this PR but I appreciate the time you spent on this.
